File: src/main/java/com/github/morningwn/t0001_t0500/t0001_t0100/t0091_t0100/T0093.java
package com.github.morningwn.t0001_t0500.t0001_t0100.t0091_t0100;

import com.github.morningwn.prefab.Out;
import org.junit.Test;

import java.util.ArrayList;
import java.util.List;

/**
 * <a href="https://leetcode.cn/problems/restore-ip-addresses/">93. 复原 IP 地址</a>
 *
 * @author morningwn
 * @create in 2023/4/1 10:49
 */
public class T0093 {

    @Test
    public void test() {
        String s = "0000";
        List<String> ipAddresses = restoreIpAddresses(s);
        Out.print(ipAddresses);
    }

    public List<String> restoreIpAddresses(String s) {
        List<String> ipAddressList = new ArrayList<>();
        getIp(new ArrayList<>(), s, 0, ipAddressList);
        return ipAddressList;
    }

    private void getIp(List<String> ipSegmentList, String s, int start, List<String> ipAddressList) {
        if (ipSegmentList.size() > 4) {
            return;
        }
        if (ipSegmentList.size() == 4) {
            if (s.length() == start) {
                ipAddressList.add(String.join(",", ipSegmentList));
            }
            return;
        }

        for (int i = 1; i < 4 && start + i <= s.length(); i++) {
            String segment = s.substring(start, start + i);
            if (check(segment)) {
                ipSegmentList.add(segment);
                getIp(ipSegmentList, s, start + i, ipAddressList);
                ipSegmentList.remove(ipSegmentList.size()-1);
            }
        }
    }

    private boolean check(String segment) {
        if (segment.length() == 1) {
            return true;
        } else if (segment.length() == 2) {
            return segment.charAt(0) != '0';
        } else if (segment.length() == 3) {
            return segment.charAt(0) != '0' && ((segment.charAt(0) - '0') * 100 + (segment.charAt(1) - '0') * 10 + (segment.charAt(2) - '0')) < 256;
        } else {
            return false;
        }
    }
}

File: src/main/go/t0001_t0500/t0001_t0100/t0091_t0100/t0093.go
package t0091_t0100

import "bytes"

// 93. 复原 IP 地址
// https://leetcode.cn/problems/restore-ip-addresses/

func restoreIpAddresses(s string) []string {
	addresses := make([]string, 0)

	for i := 1; i <= 3 && i < len(s); i++ {
		for j := i + 1; j <= i+3 && j < len(s); j++ {
			for k := j + 1; k <= j+3 && k < len(s); k++ {
				for l := k + 1; l <= k+3 && l <= len(s); l++ {
					var buffer bytes.Buffer
					if checkIpNum(s[0:i]) {
						buffer.WriteString(s[0:i])
					} else {
						continue
					}
					if checkIpNum(s[i:j]) {
						buffer.WriteString(".")
						buffer.WriteString(s[i:j])
					} else {
						continue
					}
					if checkIpNum(s[j:k]) {
						buffer.WriteString(".")
						buffer.WriteString(s[j:k])
					} else {
						continue
					}
					if checkIpNum(s[k:l]) && l == len(s) {
						buffer.WriteString(".")
						buffer.WriteString(s[k:l])
						addresses = append(addresses, buffer.String())
					}
				}
			}
		}
	}

	return addresses
}

func checkIpNum(num string) bool {
	if len(num) > 1 && num[0] == '0' {
		return false
	}
	if len(num) >= 3 && (num[0] > '2' ||
		(num[0] == '2' && num[1] > '5') ||
		(num[0] == '2' && num[1] == '5' && num[2] > '5')) {
		return false
	}
	return true
}
